design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,223 @@
+# BareMetalHost v1alpha2 API Changes
+
+## Status
+
+provisional
+
+## Summary
+
+We need to continue the API version migration for the `BareMetalHost`
+resource type to complete the work described in
+<https://github.com/metal3-io/baremetal-operator/issues/434>. This
+document lists the small API changes to be made and the steps to
+implement the changes.
+
+## Motivation
+
+The `baremetal-operator` repository still relies on a vendored copy of
+the `operator-sdk` project because the `BareMetalHost` CRD includes a
+field with type `float`, which is not formally supported by either
+`operator-sdk` or `kubebuilder`. Changing the field type will require
+a backwards-incompatible API change, so we need to change our API
+version and support upgrading existing users.
+
+### Goals
+
+- Implement a minor API version update to `v1alpha2` for the
+  `BareMetalHost` CRD.
+- Support in-place upgrades of existing `BareMetalHost` resources.
+
+### Non-Goals
+
+- Describe all possible future API changes.
+- Describe a stable `v1` API.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to update the version of metal3 in
+my existing cluster without losing any data or features.
+
+## Design Details
+
+### Choosing a New Type for the Clock Speed Field
+
+The advice from the `controller-tools` maintainers is to use
+`resource.Quantity` or `int32` for floating point values.
+
+Our problematic field holds the clock speed of a CPU in fractional
+megahertz, so in order to use an integer we would need to represent
+the speed in terms of kilohertz.
```

Comment:
I think if we represent it as an integral number of MHz, that's fine.
Originally it was an integral number of GHz, which was pretty low-resolution. I first changed it to a float, but then later changed it to MHz. In retrospect I should have changed it back to an int at that point.

Reply:
My host reports fractional MHz even for GHz. Perhaps those are representation errors (the numbers aren't identical, so that's fun).

```
$ grep 'cpu MHz' /proc/cpuinfo
cpu MHz		: 2585.951
cpu MHz		: 2399.982
cpu MHz		: 2469.934
cpu MHz		: 2400.288
cpu MHz		: 2468.399
cpu MHz		: 2417.853
cpu MHz		: 2496.642
cpu MHz		: 2412.698
cpu MHz		: 2449.773
cpu MHz		: 2399.325
cpu MHz		: 2542.550
cpu MHz		: 2399.882
cpu MHz		: 2578.015
cpu MHz		: 2399.956
cpu MHz		: 2606.565
cpu MHz		: 2404.802
cpu MHz		: 2552.326
cpu MHz		: 2412.515
cpu MHz		: 2556.837
cpu MHz		: 2400.665
cpu MHz		: 2598.182
cpu MHz		: 2400.156
cpu MHz		: 2630.709
cpu MHz		: 2412.368
cpu MHz		: 2535.367
cpu MHz		: 2400.012
cpu MHz		: 2446.512
cpu MHz		: 2402.527
cpu MHz		: 2401.146
cpu MHz		: 2412.242
cpu MHz		: 2542.022
cpu MHz		: 2412.191
cpu MHz		: 2415.755
cpu MHz		: 2399.906
cpu MHz		: 2566.748
cpu MHz		: 2399.979
cpu MHz		: 2583.524
cpu MHz		: 2401.964
cpu MHz		: 2543.863
cpu MHz		: 2409.357
```

Reply:
I suppose we could just truncate the values and turn `2409.357` to `2409`? That's still in MHz though.

Reply:
The number that IPA reports is actually the `CPU max MHz` from `lscpu`. It does report it to 4 decimal places, but I would hazard a guess that this is ~always an integer on real systems.

Annoyingly, thanks to turbo boost even this is not necessarily the actual nominal clock speed:

```
Model name:          Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
Stepping:            1
CPU MHz:             1265.260
CPU max MHz:         3100.0000
CPU min MHz:         1200.0000
```
```
Model name:                      Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
Stepping:                        10
CPU MHz:                         800.032
CPU max MHz:                     4200.0000
CPU min MHz:                     400.0000
```

The output in `/proc/cpuinfo` is just measuring noise in the dynamic frequency scaling to 3 decimal places.

Reply:
OK, so it looks like we could use an int32 with units of MHz and just truncate the float we get, turning `4200.0000` into `4200`. We still receive a string from gophercloud, so we would need to do the conversion but we're already doing the conversion to float so that's not a huge change and it will save consumers from having to do their own type conversion.